File: prodigy/sdk/rust/examples/opinionated_aegis_roundtrip.rs
// Copyright 2026 Victor Stewart
// SPDX-License-Identifier: Apache-2.0

use prodigy_sdk::aegis::AegisSession;
use prodigy_sdk::opinionated::{ActivationAction, AegisStream, PairingBook};
use prodigy_sdk::{ContainerParameters, SubscriptionPairing, U128};

fn main() -> Result<(), Box<dyn std::error::Error>>
{
   let pairing = SubscriptionPairing {
      secret: U128 {
         bytes: [
            0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d,
            0x1e, 0x1f,
         ],
      },
      address: U128 {
         bytes: [
            0xfd, 0x42, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x01,
         ],
      },
      service: 0x2233_0000_0000_1001,
      port: 3210,
      application_id: 0x2233,
      activate: true,
   };

   let parameters = ContainerParameters {
      subscription_pairings: vec![pairing],
      ..ContainerParameters::default()
   };

   let mut book = PairingBook::default();
   for action in book.seed_from_parameters(&parameters)
   {
      match action
      {
         ActivationAction::ConnectSubscriber(pairing) =>
         {
            let session = AegisSession::from_subscription(&pairing);
            let mut writer = AegisStream::with_session(session);
            let mut reader = AegisStream::with_session(session);

            let tfo_data = writer.build_tfo_data(b"mesh-aegis");
            println!("pairing_hash={:#x} tfo_bytes={}", writer.pairing_hash(), tfo_data.len());

            writer.queue_encrypted(b"ping from prodigy-sdk-opinionated")?;
            let outbound = writer.take_pending_buffer();
            reader.push_inbound(&outbound)?;
            let inbound = reader.next_inbound_message()?.expect("missing inbound message");

            println!("{}", String::from_utf8(inbound.plaintext)?);
         }
         other =>
         {
            println!("ignored startup action: {other:?}");
         }
      }
   }

   Ok(())
}

File: prodigy/sdk/cpp/examples/opinionated_aegis_roundtrip.cpp
// Copyright 2026 Victor Stewart
// SPDX-License-Identifier: Apache-2.0

#include "../opinionated/aegis_stream.h"
#include "../opinionated/pairings.h"

#include <array>
#include <cstdint>
#include <iostream>
#include <string>

namespace
{
   ProdigySDK::SubscriptionPairing demoPairing(void)
   {
      ProdigySDK::SubscriptionPairing pairing;
      pairing.secret = ProdigySDK::U128 {
         0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
         0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
      };
      pairing.address = ProdigySDK::U128 {
         0xfd, 0x42, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
      };
      pairing.service = 0x2233000000001001ULL;
      pairing.port = 3210;
      pairing.applicationID = 0x2233;
      pairing.activate = true;
      return pairing;
   }
}

int main(void)
{
   ProdigySDK::ContainerParameters parameters;
   parameters.subscriptionPairings.push_back(demoPairing());

   ProdigySDK::Opinionated::PairingBook book;
   for (const ProdigySDK::Opinionated::ActivationAction& action : book.seedFromParameters(parameters))
   {
      if (action.kind != ProdigySDK::Opinionated::ActivationActionKind::connectSubscriber ||
          action.subscription.has_value() == false)
      {
         continue;
      }

      ProdigySDK::Opinionated::AegisStream writer(*action.subscription);
      ProdigySDK::Opinionated::AegisStream reader(*action.subscription);

      const auto tfoData = writer.tfoData(
         std::span<const std::uint8_t>(
            reinterpret_cast<const std::uint8_t *>("mesh-aegis"),
            sizeof("mesh-aegis") - 1));
      std::cout << "pairing_hash=0x" << std::hex << writer.pairingHash()
                << std::dec << " tfo_bytes=" << tfoData.size() << std::endl;

      const std::array<std::uint8_t, 34> plaintext {
         'p', 'i', 'n', 'g', ' ', 'f', 'r', 'o', 'm', ' ',
         'p', 'r', 'o', 'd', 'i', 'g', 'y', '-', 's', 'd',
         'k', '-', 'o', 'p', 'i', 'n', 'i', 'o', 'n', 'a',
         't', 'e', 'd', '!'
      };
      if (writer.queueEncrypted(plaintext) == false)
      {
         std::cerr << "failed to queue outbound aegis frame" << std::endl;
         return 1;
      }

      reader.appendInbound(writer.queuedOutboundBytesView());
      if (reader.stampQueuedInboundMessages() == false)
      {
         std::cerr << "failed to stamp inbound aegis frame" << std::endl;
         return 1;
      }

      std::optional<ProdigySDK::Bytes> inbound = reader.dequeueInboundPlaintext();
      if (inbound.has_value() == false)
      {
         std::cerr << "missing decrypted plaintext" << std::endl;
         return 1;
      }

      std::cout.write(
         reinterpret_cast<const char *>(inbound->data()),
         static_cast<std::streamsize>(inbound->size()));
      std::cout << std::endl;
      return 0;
   }

   std::cerr << "missing connect-subscriber activation" << std::endl;
   return 1;
}
